lib/_007/Val.pm

```diff
@@ -205,10 +205,86 @@ class Val::Str does Val {
 ###     say(/"Bond"/.search("J. Bond"));        # --> `True`
 ###
 class Val::Regex does Val {
-    has Val::Str $.contents;
+    # note: a regex should probably keep its lexpad or something to resolve calls&identifiers
+    has $.contents;
 
     method quoted-Str {
-        "/" ~ $.contents.quoted-Str ~ "/"
+        #"/" ~ $.contents.quoted-Str ~ "/"
+        "/...a regex.../"
+    }
+
+    method match(Str $str, Bool $full) {
```

Comment:
This looks promising. I'm going to come back to reviewing it when I'm less tired.

My one comment right now &mdash; mostly unsubstantiated because I haven't taken in the whole method &mdash; is that boolean flags are generally not optimal, in the sense that there's generally a preferable alternative to them.

Reply:
Based on a full reading of `method match`, and especially what I write below about scanning, I think `match` should be three routines:

* `posmatch` &mdash; name negotiable &mdash; which attempts to make a successful match of `Str $str` starting at position `Int $pos` &mdash; returns an `Int` (with the end position) if the match was successful, and `Nil` if it wasn't. this routine can contain all the case-matching that `parse` now contains.
* `fullmatch`, which calls `posmatch($str, 0)` and then verifies that the resulting position is `$str.chars`.
* `search` which loops over all the positions in the string and returns `True` iff there is a successful `posmatch($str, $i)`.

Reply:
As a further data point, I started in on #177, and realized that `posmatch` &mdash; name negotiable &mdash; is exactly what I would need to be able to implement the grammar-matching part of `is parsed`.